### DISCUSSION TITLE
Please document the callback GIL semantics

### DISCUSSION THREAD

**Discussion by seerdecker** [pybind/pybind11#3532] [Q&A]
The doc, the functional.h file, and the examples do not specify the behavior of callbacks with respect to the GIL.

I have a C++ class that calls a std::function callback bound to a Python function from an arbitrary C++ thread asynchronously (e.g. from a C++ thread pool). The main Python thread can be anywhere during that time, e.g. it may be sleeping, it may be calling the C++ class methods with or without py::call_guard<py::gil_scoped_release>(), etc.

What happens under the hood during the callback call with respect to the GIL? How must I write the code to ensure that it's safe to call the callback? Right now I'm assuming that Pybind11 will somehow figure out whether it needs to re-acquire the GIL by magic.

Thank you

**Comment by AlexanderSerov** [comment]
Looks like satisfied [here ](https://stackoverflow.com/questions/72876146/handling-gil-when-calling-python-lambda-from-c-function) and [here](https://github.com/pybind/pybind11/pull/4057)
